slsim/Sources/QuasarCatalog/number_density_quasar.py

```diff
@@ -0,0 +1,224 @@
+    :type M_values: numpy.ndarray
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: numpy.ndarray
+
+    Returns:
+    :rtype: dict
+    :return: Dictionary containing the sorted M values and corresponding cumulative probabilities for each redshift.
+    """
+    cdf_data_dict = {}
+
+    for random_redshift in random_redshift_values:
+        dPhi_dM_values = [dPhi_dM(M, random_redshift) for M in M_values]
+        sorted_M_values = np.sort(M_values)
+        cumulative_probabilities = np.cumsum(dPhi_dM_values)
+        cumulative_prob_norm = cumulative_probabilities / max(cumulative_probabilities)
+        cdf_data_dict[random_redshift] = (sorted_M_values, cumulative_prob_norm)
+
+    return cdf_data_dict
+
+
+def cdf_fits_for_redshifts(M_values, random_redshift_values):
+    """Creates Cumulative Distribution Functions (CDF) plots for each randomly generated
+    redshift.
+
+    Parameters:
+    :param M_values: Array of absolute i-band magnitudes.
+    :type M_values: numpy.ndarray
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: numpy.ndarray
+
+    Returns:
+    :rtype: list
+    :return: List of legend handles for each CDF plot.
+    """
+    # Compute the CDF data
+    cdf_data = compute_cdf_data(M_values, random_redshift_values)
+
+    # Create an empty list to collect artists for the legend
+    legend_handles_cdf = []
+
+    # Plot the CDF vs. M curve for each randomly generated redshift
+    for random_redshift, (sorted_M_values, cumulative_prob_norm) in cdf_data.items():
+        (line,) = plt.plot(
+            sorted_M_values, cumulative_prob_norm, label=f"z={random_redshift}"
+        )
+        legend_handles_cdf.append(line)
+
+    return legend_handles_cdf
+
+
+def inverse_cdf_fits_for_redshifts(M_values, random_redshift_values):
+    """Creates inverse Cumulative Distribution Function (CDF) fits for each randomly
+    generated redshift.
+
+    Parameters:
+    :param M_values: Array of absolute i-band magnitudes.
+    :type M_values: numpy.ndarray
+    :param random_redshift_values: Array of randomly generated redshift values.
+    :type random_redshift_values: numpy.ndarray
+
+    Returns:
+    :rtype: dict
+    :return: Dictionary of inverse CDF functions for each redshift.
+    """
+    # Compute the CDF data
+    cdf_data = compute_cdf_data(M_values, random_redshift_values)
+
+    inverse_cdf_dict = {}
+
+    for random_redshift, (sorted_M_values, cumulative_prob_norm) in cdf_data.items():
+        # Interpolate the inverse CDF function
+        inverse_cdf = interp1d(
+            cumulative_prob_norm,
+            sorted_M_values,
+            kind="linear",
+            fill_value="extrapolate",
+        )
+        inverse_cdf_dict[random_redshift] = inverse_cdf
+
+    return inverse_cdf_dict
+
+
+def generate_redshift_table(random_redshift_values, inverse_cdf_dict, seed=42):
```

Comment:
You do not need inverse_cdf_dict if you call inverse_cdf_fits_for_redshifts() function inside this function. see comment below.